### PR TITLE
AVL Tree in C++

### PR DESCRIPTION
**Pull request by tanya12181**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?
Yes

### Description

I have added the implementation of AVL trees in c plus plus/ds/ folder. Kindly review my PR @bhav09.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Pull request for issue (https://github.com/TesseractCoding/NeoAlgo/issues/2696): Implementation of AVL trees in C++.

### CODE REVIEW

File: C-Plus-Plus/ds/avl_tree.cpp [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+                   /  \     / \
+                  4   11   16  60
+                     /       \
+                    8         20
+
+ */
+//Other cases that you could try:
+//vector<int>data={10,20,30,40,50,25};
+//vector<int>data={33,13,53,9,21,61,8,11};
+//vector<int>data={14,13};
```
Comment (bhav09) [inline C-Plus-Plus/ds/avl_tree.cpp:540]: Add a new line here.

File: C-Plus-Plus/ds/avl_tree.cpp [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    for(int i=0;i<deletion_order.size();i++){
+        cin>>deletion_order[i];
+        root=delete_node(root,deletion_order[i],root->data);
+        cout<<"\nafter deleting node with key value = "<<deletion_order[i]<<"\n";
+        preorder_traversal(root);
+        cout<<"\n";
+    }
+    return 0;
+}
+/*
```
Comment (bhav09) [inline C-Plus-Plus/ds/avl_tree.cpp:457]: Need to add complexity.

File: C-Plus-Plus/ds/avl_tree.cpp [hunk 3]
```diff
@@ -0,0 +1,481 @@
+#include <iostream>
```
Comment (bhav09) [inline C-Plus-Plus/ds/avl_tree.cpp:11]: Describe the AVL tree in brief., here.

File: C-Plus-Plus/ds/avl_tree.cpp [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+    Node* next = critical_node->left;
+    Node* temp = next->right;
+    next->right=critical_node;
+    critical_node->left=temp;
+
+    //updating height and balance factors for the rotated nodes
+    next->height=find_height(next);
+    next->bf = find_balance_factor(next);
+
+    critical_node->height=find_height(critical_node);
```
Comment (bhav09) [inline C-Plus-Plus/ds/avl_tree.cpp:82]: The code lacks comments. Please add comments to it. Comments should above the code segment and not alongside.

File: C-Plus-Plus/ds/avl_tree.cpp [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+        }
+        else if(root->data<node->data){
+            root=root->right;
+        }
+    }
+    return(parent);
+}
+
+/*
+This function finds the critical node in the AVL Tree
```
Comment (bhav09) [inline C-Plus-Plus/ds/avl_tree.cpp:140]: Good.

File: C-Plus-Plus/ds/avl_tree.cpp [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+    node=node->left;
+    while(node->right!=NULL){
+        parent=node;
+        node=node->right;
+    }
+    return(node);
+}
+
+/*
+this function balances our AVL tree(in case of deletion).
```
Comment (bhav09) [inline C-Plus-Plus/ds/avl_tree.cpp:187]: Good

File: C-Plus-Plus/ds/avl_tree.cpp (hunk 1 above)
Comment (tanya12181) [reply C-Plus-Plus/ds/avl_tree.cpp:540]: sir, I am not sure I understand. Are you saying to add a blank line after "Other cases that you could try:" ?

Comment (bhav09) [reply C-Plus-Plus/ds/avl_tree.cpp:540]: Add a blank at the end of the file.

**Comment by tanya12181** [comment]
@bhav09  sir, I have resolved the issues in my PR. please review it

**Comment by bhav09** [comment]
Great work, Tanya!

**Comment by HarshCasper** [comment]
@all-contributors bot please add @tanya12181 for code

**Pull request merged**